### PR TITLE
Add support for CustomAutoRooter

### PR DESCRIPTION
**Pull request by michaelwu**
As requested by @jdm . Fixes https://github.com/servo/servo/issues/7676. Not particularly great to use but it works for me.

[Review on Reviewable](https://reviewable.io/reviews/servo/rust-mozjs/225)

### CODE REVIEW

File: src/rust.rs [hunk 1]
```diff
@@ -336,6 +337,40 @@ impl<T> Drop for Rooted<T> {
...
 
+impl CustomAutoRooter {
+    pub fn new(cx: *mut JSContext, vftable: &'static _vftable_CustomAutoRooter)
+        -> CustomAutoRooter {
+        CustomAutoRooter::new_with_addr(cx, vftable, unsafe { return_address() })
+    }
+
+    pub fn new_with_addr(cx: *mut JSContext, vftable: &'static _vftable_CustomAutoRooter, addr: *const u8) -> CustomAutoRooter {
+        let ctxfriend: &mut ContextFriendFields = unsafe {
+            mem::transmute(cx)
```
Comment (jdm) [inline src/rust.rs]: I would prefer `&mut *(cx as *mut ContextFriendFields)` instead of a transmute.

File: src/rust.rs [hunk 2]
```diff
@@ -336,6 +337,40 @@ impl<T> Drop for Rooted<T> {
...
+        let rooter = CustomAutoRooter {
+            _vftable: vftable as *const _,
+            _base: AutoGCRooter {
+                down: ctxfriend.autoGCRooters,
+                tag_: AutoGCRooter_enum0::CUSTOM as ptrdiff_t,
+                stackTop: &mut ctxfriend.autoGCRooters as *mut _,
+            }
+        };
+
+        ctxfriend.autoGCRooters = unsafe { mem::transmute((addr as *const *const u8).offset(1)) };
```
Comment (jdm) [inline src/rust.rs]: I'd prefer explicit casts here instead of transmute.

File: src/rust.rs [hunk 3]
```diff
@@ -336,6 +337,40 @@ impl<T> Drop for Rooted<T> {
...
+        rooter
+    }
+}
+
+impl Drop for CustomAutoRooter {
```
Comment (jdm) [inline src/rust.rs]: Bleh, another destructor on a repr(C) type :/

**Comment by jdm** [comment]
I'd like to see the transmutes replaced with casts; otherwise this looks ok.

**Comment by michaelwu** [comment]
Transmutes replaced with casts.

**Pull request merged**